File: language/src/main/java/brj/nodes/DefRootNode.java
package brj.nodes;

import brj.BridjeContext;
import brj.BridjeLanguage;
import brj.Typing;
import brj.runtime.Symbol;
import com.oracle.truffle.api.dsl.CachedContext;
import com.oracle.truffle.api.dsl.NodeChild;
import com.oracle.truffle.api.dsl.NodeField;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.frame.FrameDescriptor;
import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.nodes.RootNode;
import com.oracle.truffle.api.source.SourceSection;
import org.jetbrains.annotations.NotNull;

@NodeField(name = "sym", type = Symbol.class)
@NodeField(name = "typing", type = Typing.class)
@NodeChild(value = "expr", type = ExprNode.class)
@NodeField(name = "sourceSection", type = SourceSection.class)
public abstract class DefRootNode extends RootNode {

    protected DefRootNode(BridjeLanguage lang, FrameDescriptor frameDescriptor) {
        super(lang, frameDescriptor);
    }

    public abstract Symbol getSym();

    public abstract Typing getTyping();

    @NotNull
    @Specialization
    public Object doExecute(VirtualFrame frame,
                            Object exprVal,
                            @CachedContext(BridjeLanguage.class) BridjeContext ctx) {
        ctx.getBridjeEnv().def(getSym(), getTyping(), exprVal);
        return exprVal;
    }
}

File: language/src/main/kotlin/brj/nodes/DefRootNode.kt
package brj.nodes

import brj.BridjeLanguage
import brj.Typing
import brj.runtime.NsContext
import brj.runtime.Symbol
import com.oracle.truffle.api.dsl.NodeChild
import com.oracle.truffle.api.dsl.Specialization
import com.oracle.truffle.api.frame.FrameDescriptor
import com.oracle.truffle.api.nodes.RootNode
import com.oracle.truffle.api.source.SourceSection

@NodeChild(value = "expr", type = ExprNode::class)
internal abstract class DefRootNode(
    lang: BridjeLanguage, frameDescriptor: FrameDescriptor,
    private val nsCtx: NsContext,
    private val sym: Symbol,
    private val typing: Typing,
    private val loc: SourceSection?
) : RootNode(lang, frameDescriptor) {

    @Specialization
    fun doExecute(exprVal: Any): Any {
        nsCtx.def(sym, typing, exprVal)
        return exprVal
    }

    override fun getSourceSection() = loc
}